### PR TITLE
Add HeavenManga[ES] extension

### LINKED ISSUE #1335: [Source Request][Spanish] HeavenManga
Source: http://heavenmanga.com/

### PR DESCRIPTION
**Pull request by waptik**
Closes #1335

### CODE REVIEW

**Comment by waptik** [comment]
After submitting the PR i just found out some previous commits of another branch(manmanga) were included in my push. I must have missed them. Sorry.
But they won't affect manmanga extension.

**Comment by arkon** [comment]
Doesn't the repo already contain the manmanga changes? Can you rebase/update your branch?

**Comment by waptik** [comment]
> Doesn't the repo already contain the manmanga changes? Can you rebase/update your branch?

Done by rebasing it!

File: src/es/heavenmanga/build.gradle [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ext {
+    appName = 'Tachiyomi: HeavenManga'
+    pkgNameSuffix = 'es.heavenmanga'
+    extClass = '.HeavenManga'
+    extVersionCode = 1
+    libVersion = '1.2'
+}
+
+dependencies {
+    implementation project(':lib-ratelimit')
```
Comment (arkon) [inline src/es/heavenmanga/build.gradle]: Don't need this.

File: src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+        }
+
+        return GET(search_url, headers)
+    }
+
+    override fun imageUrlRequest(page: Page) = GET(page.url, headers)
+
+    // get contents of a url
+    private fun getUrlContents(url: String): Document = Jsoup.connect(url).timeout(0).get()
```
Comment (arkon) [inline src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt]: This is a weird way of doing it. I think you can find examples of manually calling `GET(...)` synchronously in some other extensions.

File: src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+    override fun chapterFromElement(element: Element): SChapter {
+        val urlElement = element.select("a").first()
+        val timeElement = element.select("span").first()
+        val time = timeElement.text()
+        val date = time.replace("--", "-")
+        val url = urlElement.attr("href")
+
+        val chapter = SChapter.create()
+        Log.d("HM: chapUrl->", chapter.setUrlWithoutDomain(url).toString())
```
Comment (arkon) [inline src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt]: Remove the logs.

File: src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        return pages
+    }
+
+    /**
+     * Array.from(document.querySelectorAll('.categorias a')).map(a => `Pair("${a.textContent}", "${a.getAttribute('href')}")`).join(',\n')
+     * on http://heavenmanga.com/top/
+     * */
+    private class GenreFilter : UriPartFilter("Géneros", arrayOf(
+        Pair("Todo", ""),
+        Pair("Accion", "/genero/accion.html"),
```
Comment (arkon) [inline src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt]: You could refactor this to remove all of the redundant `/genero/[...].html` parts and just handle that in the search logic. Same goes for some of the other filters.

File: src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt (hunk 2 above)
Comment (waptik) [reply src/es/heavenmanga/src/eu/kanade/tachiyomi/extension/es/heavenmanga/HeavenManga.kt]: Are you talking about `private fun getUrlContents(url: String): Document = Jsoup.connect(url).timeout(0).get()` ?

**Pull request merged**